bytecomp/translcore.ml

```diff
@@ -1232,11 +1241,44 @@ and transl_record env all_labels repres lbl_expr_list opt_init_expr =
     end
   end
 
-and transl_match e arg pat_expr_list exn_pat_expr_list partial =
-  let id = name_pattern "exn" exn_pat_expr_list
-  and cases = transl_cases pat_expr_list
-  and exn_cases = transl_cases exn_pat_expr_list in
+and transl_match e arg pat_expr_list partial =
+  let rewrite_case (val_cases, exn_cases, static_handlers as acc)
+        ({ c_lhs; c_guard; c_rhs } as case) =
+    if c_rhs.exp_desc = Texp_unreachable then acc else
+    let val_pat, exn_pat = split_pattern c_lhs in
+    match val_pat, exn_pat with
+    | None, None -> assert false
+    | Some pv, None ->
+        assert (Typedtree.pat_equiv pv c_lhs);
+        let val_case = transl_case { case with c_lhs = pv } in
+        val_case :: val_cases, exn_cases, static_handlers
+    | None, Some pe ->
+        let exn_case = transl_case { case with c_lhs = pe } in
+        val_cases, exn_case :: exn_cases, static_handlers
+    | Some pv, Some pe ->
+        assert (c_guard = None);
+        let ids  = Typedtree.pat_bound_idents pv in
+        let vars = List.map (fun id -> Lvar id) ids in
+        let lbl  = next_negative_raise_count () in
+        let body = Lstaticraise (lbl, vars) in
+        (* Simplif doesn't like it if binders are not uniq, so we make sure to
+           use different names in the value and the exception branches. *)
```

Comment:
Oh, indeed. But then this could be clarified, with a comment pointing that out, but maybe even better by writing the code in a way that looks less like a whole-expression duplication. Maybe something like

```ocaml
  let static_raise ids = Lstaticraise (lbl, List.map (fun id -> Lvar id) ids) in
  (* blabla avoid duplicate binders *)
  let fresh_ids = ... in
  let fresh_pe = alpha_pat ... pe in
  (pv, static_raise ids) :: val_cases,
  (fresh_pe, static_raise fresh_ids) :: exn_cases
```